notebooks/generative/fashionmnist_vae.livemd

```diff
@@ -0,0 +1,304 @@
+  """
+  def preprocess_data(data) do
+    {image_data, _labels} = data
+    {images_binary, type, shape} = image_data
+
+    images_binary
+    |> Nx.from_binary(type)
+    # Since pixels are organized row-wise, reshape into rows x columns
+    |> Nx.reshape(shape, names: [:images, :channels, :height, :width])
+    # Normalize the pixel values to be between 0 and 1
+    |> Nx.divide(255)
+  end
+
+  @doc """
+    Converts a tensor of images into paired, batched images for model training
+  """
+  def prepare_training_data(images, batch_size) do
+    batched_images = images |> Nx.to_batched(batch_size)
+    Stream.zip(batched_images, batched_images)
+  end
+end
+```
+
+```elixir
+train_images = Utils.preprocess_data(Scidata.FashionMNIST.download())
+test_images = Utils.preprocess_data(Scidata.FashionMNIST.download_test())
+
+Kino.render(train_images[[images: 0]] |> Utils.image_to_kino())
+Kino.render(test_images[[images: 0]] |> Utils.image_to_kino())
+
+:ok
+```
+
+Our simple autoencoder model. We won't be using a denoising autoencoder here.
+
+Note that we're giving each of the layers a name - the reason for this will be apparent later.
+
+```elixir
+model =
+  Axon.input("image", shape: {nil, 1, 28, 28})
+  # This is now 28*28*1 = 784
+  |> Axon.flatten()
+  # The encoder
+  |> Axon.dense(256, activation: :relu, name: "encoder_layer_1")
+  |> Axon.dense(128, activation: :relu, name: "encoder_layer_2")
+  |> Axon.dense(64, activation: :relu, name: "encoder_layer_3")
+  # Bottleneck layer
+  |> Axon.dense(10, activation: :relu, name: "bottleneck_layer")
+  # The decoder
+  |> Axon.dense(64, activation: :relu, name: "decoder_layer_1")
+  |> Axon.dense(128, activation: :relu, name: "decoder_layer_2")
+  |> Axon.dense(256, activation: :relu, name: "decoder_layer_3")
+  |> Axon.dense(784, activation: :sigmoid, name: "decoder_layer_4")
+  # Turn it back into a 28x28 single channel image
+  |> Axon.reshape({1, 28, 28})
+
+# We can use Axon.Display to show us what each of the layers would look like
+# assuming we send in a batch of 4 images
+Axon.Display.as_table(model, Nx.template({4, 1, 28, 28}, :f32)) |> IO.puts()
+```
+
+```elixir
+batch_size = 128
+
+train_data = Utils.prepare_training_data(train_images, 128)
+test_data = Utils.prepare_training_data(test_images, 128)
+
+{input_batch, target_batch} = Enum.at(train_data, 0)
+Kino.render(input_batch[[images: 0]] |> Utils.image_to_kino())
+Kino.render(target_batch[[images: 0]] |> Utils.image_to_kino())
+
+:ok
+```
+
+When training, it can be useful to stop execution early - either when you see it's failing and you don't want to waste time waiting for the remaining epochs to finish, or if it's good enough and you want to start experimenting with it.
+
+The `livebook_early_stop/1` function below is a handy handler to give us a `Kino.Control.button` that will stop the training loop when clicked.
+
+```elixir
+defmodule LivebookTraining do
+  def livebook_early_stop(loop) do
```

Comment:
This is **really** cool and definitely belongs in a library somewhere. If you want to open a PR to Axon I would gladly accept it, but we are also planning on a Kino-Axon specific library, so perhaps it's best that lives there. Either way I think opening a PR in Axon is the best thing to do right now so we remember to extract this out when Kino-Axon integration comes to life :)

Reply:
I was thinking the same thing - a `Kino-Axon` library would be the perfect place for it. I'll create a PR in the meantime, as soon as I figure out how to write a test for it :)

Reply:
Don't let the test hold you back from a PR, we can figure something out during the review!